### ISSUE TITLE
Improve command line experience.

### ISSUE THREAD

**Issue by purpleP** [neovim/neovim#5604]
One of the things that vim/neovim could improve is command completion so that user wouldn't press tab to see options. We all love YouCompleteMe, deoplete etc, right? Why do we still have to press tab in vim command mode as if we are in the previous century?

It so happens that I've seen emacs ido plugin at work, and I liked its approach so much, that I've started learning emacs a bit. You can go find videos on youtube about it, but what it is does essentially is showing available options **horizontally** and resorts them **as you type** while **highlighting** what you've typed so far.

This what I wanted vim command line to look like. And it's not just me, there's is at least this [guy](https://github.com/paradigm/SkyBison) and Unite/Denite menus.

People have made a lot of work, to make workarounds around this limitation and reimplemented a lot of what vim can do out of the box, because of that limitation (Think about Ctrl-p, Command-t, Unite, Denite etc). And while they can do more, as they can show more options because of their vertical layout, it's also their limitation for me (and many more people that are using ido instead of HELM in emacs).

So what limitation am I talking about? Somehow big vertical buffer with a lot of options distracts my flow of concentration. I think main reasons is that you don't know where the top of it will be exactly and when I'm pressing `<C-p>` it's like a little have bomb just exploded in my buffer and I need to find the top line and skim over options and only after that I can type. This messes my concentration on code somehow, while using `:find` and pressing tabs and seeing options right there on the same line I type is much less distracting.

So what I like to see in vim is something like that. starts means highlighting.
`:**b**{ b | bd | bufdo }`
Then say I want to switch to other buffer so I pressed space 
`:**b **{ foo.txt, bar.txt}`
I want `bar.txt` so I press `b`
`:**b b**{ bar.txt}` or maybe `:b **b**{ bar.txt}` meaning the options that is already fixed shouldn't be highlighted.
I press enter here and vim opens file that I wanted.

This should be simple to implement, because vim can already autocomplete in commands It just need to do that automatically and to be able to fuzzy matching. If the autocomplete is taking to long it can show `{...} or something and autocomplete should be done async, so that user can type without waiting for options.

I believe this thing is very important, because it would make vanilla vim much more pleasant to use, and less plugins pretty much always the better.

One other thing that I'd like to see in vim command line is the ability to hide it completely when it's not needed. Right now it always eat one line at the bottom, even if user isn't typing any command.

The command line could be made pluggable somehow and what I'm talking about could be just plugin that is installed by default like netrw, or this can be just options and behaviour that is added directly to vim/neovim code base.

I'm willing to work on this features, but I would need some guidance. I've read readme and the project root, but haven't able to find where is the code that deals with command line is.
My C knowledge is incomplete and rusty, but I'm always willing to learn.

**Comment by justinmk** [comment]
> many more people that are using ido instead of HELM in emacs

This isn't a strong argument to me. fzf and HELM are objectively better than ido, even after hacking up ido's fuzzy-matching. 

> One other thing that I'd like to see in vim command line is the ability to hide it completely when it's not needed.

https://github.com/neovim/neovim/issues/1004

> The command line could be made pluggable somehow and what I'm talking about could be just plugin that is installed by default like netrw, 

That's probably a good direction. It will take some work.

> where is the code that deals with command line is.

`ex_getln.c`

Definitely, if you want to work on this, and the work has test coverage and documentation, there's nothing standing in the way of getting such a change merged. 

However let's merge this issue into https://github.com/neovim/neovim/issues/3013, closing.

**Comment by purpleP** [comment]
@justinmk 
> This isn't a strong argument to me.
What isn't strong about it? There are people that like Ido more than helm, there are people that like helm more than Ido. It doesn't matter what you personally like. What matter is that there a a lot of people that would like this.
Probably all you've said actually is that you like helm more and it's fine. what I'm saying is that people should have choice.

**Comment by justinmk** [comment]
Sure. I said other things in that comment too :)

**Comment by purpleP** [comment]
@justinmk Yes you did and I agree with them.

**Comment by trusktr** [comment]
> This isn't a strong argument to me. fzf and HELM are objectively better than ido,

Just curious: how do we augment command line with fzf? Suppose I type `:fi`. Normally I'd hit `<tab>` and get (neo)vim's completion menu above the command line. What would I do to trigger `fzf` to complete that command instead? Can I autocomplete stuff that's in my buffers, etc, just like I normally do with `<c-n>` while in INSERT mode?


It'd be nice to have a vertical completion list (like `<c-n>` in INSERT mode) with fuzzy abilities (like fzf). As far as I've known, I haven't seen this.

For example, to use `fzf`, I can use the `:Commands` command to fuzzy find commands, but once I've chosen a command, fzf is gone, and then completing any arguments returns back to the less-flexible builtin completion.


I think this issue is a good one. I think @purpleP just wants better completion like fzf or ido builtin, completely, not half way (like fzf is half-way, unless I've missed something).

**Comment by purpleP** [comment]
@trusktr

Yes, that's exactly what I want.

**Comment by bfredl** [comment]
I suppose what is needed is something like `complete()` but works in cmdline mode instead of insert mode (I think even same function can be used, if not some too popular plugin speculatively calls it at any time without checking mode)

Also there is really no big roadblock of using the vertical completion popupmenu in cmdline mode. It's implementation is quite isolated from insert mode specific completion logic. I think reason horizontal wildmenu was implemented was to avoid redrawing in cmdline mode (notice how wildmenu is never cleared, only overdrawn by another menu). But given how much useful stuff that now is done that does redraw in cmdline mode (inccommand, improved incsearch etc, not to mention vital-over based plugins who could use the "native" cmdline if it was improved), I think "avoid redraws in cmdline mode" is a prior to be dropped.
